aggregator/src/cache.rs

```diff
@@ -206,3 +207,240 @@ impl PeerAggregatorCache {
             .find(|peer| peer.endpoint() == endpoint && peer.role() == role)
     }
 }
+
+#[derive(Debug)]
+pub struct TaskAggregatorCache<C: Clock> {
+    datastore: Arc<Datastore<C>>,
+    report_writer: Arc<ReportWriteBatcher<C>>,
+    cache: Cache<TaskId, TaskAggregatorCacheEntry<C>>,
+    cache_none: bool,
+}
+
+/// A cache entry. None indicates that the entry has not been initialized yet.
+type TaskAggregatorCacheEntry<C> = Arc<Mutex<Option<TaskAggregatorRef<C>>>>;
+
+/// An Arc reference to a TaskAggregator. None indicates that there is no such task aggregator in
+/// the database.
+type TaskAggregatorRef<C> = Option<Arc<TaskAggregator<C>>>;
+
+pub const TASK_AGGREGATOR_CACHE_DEFAULT_TTL: Duration = Duration::from_secs(600);
+pub const TASK_AGGREGATOR_CACHE_DEFAULT_CAPACITY: u64 = 10_000;
+
+impl<C: Clock> TaskAggregatorCache<C> {
+    pub fn new(
+        datastore: Arc<Datastore<C>>,
+        report_writer: ReportWriteBatcher<C>,
+        cache_none: bool,
+        capacity: u64,
+        ttl: Duration,
+    ) -> Self {
+        Self {
+            datastore,
+            report_writer: Arc::new(report_writer),
+            cache: CacheBuilder::new(capacity).time_to_live(ttl).build(),
+            cache_none,
+        }
+    }
+
+    pub async fn get(&self, task_id: &TaskId) -> Result<TaskAggregatorRef<C>, Error> {
+        // Step one: grab the existing entry for this task, if one exists. If there is no existing
+        // entry, write a new uninitialized entry.
+        let entry = self
+            .cache
+            .get_with(*task_id, async { Default::default() })
```

Comment:
Oh! I think the approach in the second paragraph is just [`Cache::try_get_with`](https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.try_get_with). I think we can effectively move all of our initialization logic into a call to `try_get_with` -- it even returns an `Arc<Error>`, so I guess we should go with that approach.

We could even return an error on the task-does-not-exist case if `cache_none` is false to avoid ever populating a `None` in that case, as long as we translated that error back to a `None` return value from the function. (We might also do something with [`.entry().and_try_compute_with()`](https://docs.rs/moka/latest/moka/future/struct.OwnedKeyEntrySelector.html#method.and_try_compute_with) which removes the need for this error translation, but that API is pretty complicated...)

Reply:
I'm glad we went with this library and that you found that API, as the result is remarkably clean.

Reply:
`.and_try_compute_with()` does have a bit of an intimidating API, but I find the result to be easy to reason about, since manipulation of the cache is clearly articulated by `Op::Nop`, `Op::Put`, etc.

Reply:
Very nice! I was hoping we could reduce this function to a single request into the cache.